treeherder/log_parser/failureline.py

```diff
@@ -204,3 +215,468 @@ def get_group_results(repository, push):
+
+        for task_id, groups_json in cursor:
+            if groups_json:
+                # PostgreSQL json_object_agg returns a dict directly, not a JSON string
+                by_task_id[task_id] = groups_json
+
+    return by_task_id
+
+
+def get_group_results_new_direct(revision, repository_name):
+    """
+    Most optimized version - bypasses ALL ORM queries by using revision/repository directly.
+    Single SQL query handles everything including push lookup.
+    """
+    from django.db import connection
+
+    # PostgreSQL version with json_object_agg
+    query = """
+        SELECT 
+            tcm.task_id,
+            json_object_agg(
+                g.name,
+                CASE WHEN gs.status = %s THEN true ELSE false END
+            ) as groups_data
+        FROM group_status gs
+        INNER JOIN job_log jl ON gs.job_log_id = jl.id
+        INNER JOIN job j ON jl.job_id = j.id
+        INNER JOIN push p ON j.push_id = p.id
+        INNER JOIN repository r ON p.repository_id = r.id
+        INNER JOIN taskcluster_metadata tcm ON j.id = tcm.job_id
+        INNER JOIN "group" g ON gs.group_id = g.id
+        WHERE p.revision = %s
+        AND r.name = %s
+        AND gs.status IN (%s, %s)
+        GROUP BY tcm.task_id
+    """
+
+    by_task_id = {}
+
+    with connection.cursor() as cursor:
+        cursor.execute(
+            query, [GroupStatus.OK, revision, repository_name, GroupStatus.OK, GroupStatus.ERROR]
+        )
+
+        for task_id, groups_json in cursor:
+            if groups_json:
+                # PostgreSQL json_object_agg returns a dict directly, not a JSON string
+                by_task_id[task_id] = groups_json
+
+    return by_task_id
+
+
+def get_group_results_new_fast_dict(push):
+    """
+    Optimized version using regular dict for better performance.
+    Uses PostgreSQL syntax with "group" table properly quoted.
+    """
+    from django.db import connection
+
+    # Pre-compute the OK status value
+    OK_STATUS = GroupStatus.OK
+
+    query = """
+        SELECT 
+            tcm.task_id,
+            g.name,
+            gs.status
+        FROM group_status gs
+        INNER JOIN job_log jl ON gs.job_log_id = jl.id
+        INNER JOIN job j ON jl.job_id = j.id
+        INNER JOIN taskcluster_metadata tcm ON j.id = tcm.job_id
+        INNER JOIN "group" g ON gs.group_id = g.id
+        WHERE j.push_id = %s
+        AND gs.status IN (%s, %s)
+        ORDER BY tcm.task_id
+    """
+
+    by_task_id = {}
+
+    with connection.cursor() as cursor:
+        cursor.execute(query, [push.id, GroupStatus.OK, GroupStatus.ERROR])
+
+        # Fetch all results at once if dataset is manageable
+        # or use fetchmany for very large datasets
+        rows = cursor.fetchall()
+
+        # Process with regular dict - often faster than defaultdict
+        for task_id, group_name, status in rows:
+            if task_id not in by_task_id:
+                by_task_id[task_id] = {}
```

Comment:
I wonder if using defaultdict() would be more efficient, something like:
```
by_task_id = defaultdict(dict)
for task_id, group_name, status in rows:
  by_task_id[task_id][group_name] = status == OK_STATUS
```

I think it might be a wash given that the dict is built once/query